src/github.com/matrix-org/dendrite/mediaapi/writers/upload.go

```diff
@@ -0,0 +1,244 @@
+// Copyright 2017 Vector Creations Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package writers
+
+import (
+	"crypto/sha256"
+	"database/sql"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/matrix-org/dendrite/clientapi/auth"
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/dendrite/mediaapi/config"
+	"github.com/matrix-org/dendrite/mediaapi/storage"
+	"github.com/matrix-org/dendrite/mediaapi/types"
+	"github.com/matrix-org/util"
+)
+
+// uploadRequest metadata included in or derivable from an upload request
+// https://matrix.org/docs/spec/client_server/r0.2.0.html#post-matrix-media-r0-upload
+// NOTE: The members come from HTTP request metadata such as headers, query parameters or can be derived from such
+type uploadRequest struct {
+	MediaMetadata *types.MediaMetadata
+}
+
+// Validate validates the uploadRequest fields
+func (r uploadRequest) Validate(maxFileSizeBytes types.ContentLength) *util.JSONResponse {
+	// TODO: Any validation to be done on ContentDisposition?
+
+	if r.MediaMetadata.ContentLength < 1 {
+		return &util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.Unknown("HTTP Content-Length request header must be greater than zero."),
+		}
+	}
+	if maxFileSizeBytes > 0 && r.MediaMetadata.ContentLength > maxFileSizeBytes {
+		return &util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.Unknown(fmt.Sprintf("HTTP Content-Length is greater than the maximum allowed upload size (%v).", maxFileSizeBytes)),
+		}
+	}
+	// TODO: Check if the Content-Type is a valid type?
+	if r.MediaMetadata.ContentType == "" {
```

Comment:
During implementation, I considered `Validate()` to apply to incoming requests to us. As such, in the download handler, there is a `Validate()` that simply checks the Origin and MediaID are not empty strings though it should rather check that Origin is a valid and reachable domain (which it ultimately will be if we manage to get a file from it and the request would fail if not) and the MediaID is valid (that is difficult for us to know as the format of MediaIDs is not specified. Plus it would also fail if the remote server considers it an invalid MediaID and I feel we have to consider the remote server the source of truth on its MediaIDs.)